src/net.h

```diff
@@ -1267,6 +1267,36 @@ class CConnman
      */
     std::vector<CService> m_onion_binds;
 
+    /**
+     * RAII helper to atomically create a copy of `vNodes` and add a reference
+     * to each of the nodes. The nodes are released when this object is destroyed.
+     */
+    class NodesSnapshot
+    {
+    public:
+        explicit NodesSnapshot(CConnman* connman) : m_connman{connman}
+        {
+            LOCK(m_connman->cs_vNodes);
+            m_nodes_copy = m_connman->vNodes;
+            for (auto& node : m_nodes_copy) {
+                node->AddRef();
+            }
+        }
+
+        ~NodesSnapshot()
+        {
+            LOCK(m_connman->cs_vNodes);
+            for (auto& node : m_nodes_copy) {
+                node->Release();
+            }
+        }
+
+        std::vector<CNode*> m_nodes_copy;
+
+    private:
+        CConnman* m_connman;
```

Comment:
96bcc72b5dc4f8ba5654d8ddeab844688507f1ce

nit, * const

Reply:
I did though of that but then the caller inits with `{*this}`?

Reply:
After reading
https://github.com/isocpp/CppCoreGuidelines/issues/222 and
https://github.com/isocpp/CppCoreGuidelines/issues/1707
I decided to use a reference instead of a pointer because the member cannot/must not be nullable.